counsel.el

```diff
@@ -6312,6 +6327,19 @@ Use the presence of a \".git\" file to determine the root."
 Use the presence of a `dir-locals-file' to determine the root."
   (counsel--dominating-file dir-locals-file))
 
+(defun counsel--single-file-root ()
+  "Return the root of a single standalone file.
+
+To avoid catching un-related history for this one file we locally
+override `counsel-compile-history' and leave the user to fill in their
+preferred build stanza.
+
+This should be the last function in `counsel-compile-root-functions'
+after giving all the other project root finders a chance."
+  (unless (local-variable-p 'counsel-compile-history-exact-src)
+    (setq-local counsel-compile-history-exact-src 't))
+  (file-name-directory (or (buffer-file-name) default-directory)))
```

Comment:
Why not just return `default-directory` directly?

Reply:
I guess - can't default-directory get changed by other things though?

Reply:
> I guess - can't default-directory get changed by other things though?

I don't understand what you mean or what the concern is, could you please elaborate?

Reply:
Thinking on it maybe I'm thinking more of buffers which aren't backed by files. We can assume that when buffer-file-name exists default-directory will always be the same as (file-name-directory buffer-file-name)?

Reply:
In well-behaved programs, I think so.